src/mount.c

```diff
@@ -36,78 +38,163 @@ pu_create_mount_point(const gchar *name,
     return mount_point;
 }
 
+static GPtrArray *
+pu_find_mounted(const gchar *device)
+{
+    blkid_probe pr;
+    blkid_partlist ls;
+    gint i;
+    gint num_partitions;
+    struct libmnt_table *tb;
+    gchar *partname;
+    g_autoptr(GPtrArray) mounted = g_ptr_array_new_with_free_func(g_free);
+    g_autoptr(GError) error = NULL;
+
+    g_return_val_if_fail(g_strcmp0(device, "") > 0, NULL);
+
+    pr = blkid_new_probe_from_filename(device);
+    if (!pr)
+        return NULL;
+
+    ls = blkid_probe_get_partitions(pr);
+    if (!ls) {
+        blkid_free_probe(pr);
+        return NULL;
+    }
+    num_partitions = blkid_partlist_numof_partitions(ls);
+    if (num_partitions < 0) {
+        blkid_free_probe(pr);
+        return NULL;
+    }
+    blkid_free_probe(pr);
+
+    tb = mnt_new_table();
+    if (!tb)
+        return NULL;
+
+    if (mnt_table_parse_mtab(tb, NULL) < 0) {
+        mnt_unref_table(tb);
+        return NULL;
+    }
+
+    for (i = 0; i < num_partitions; i++) {
+        partname = pu_device_get_partition_path(device, i + 1, &error);
+        if (!partname) {
+            mnt_unref_table(tb);
+            return NULL;
+        }
+        if (mnt_table_find_source(tb, partname, MNT_ITER_FORWARD))
+            g_ptr_array_add(mounted, partname);
+    }
+
+    mnt_unref_table(tb);
+    return g_steal_pointer(&mounted);
+}
+
 gboolean
 pu_mount(const gchar *source,
          const gchar *mount_point,
          GError **error)
 {
-    g_autofree gchar *cmd = g_strdup_printf("mount %s %s", source, mount_point);
+    gint ret;
+    struct libmnt_context *cxt;
 
+    g_return_val_if_fail(g_strcmp0(source, "") > 0, FALSE);
+    g_return_val_if_fail(g_strcmp0(mount_point, "") > 0, FALSE);
     g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
 
-    if (!g_spawn_command_line_sync(cmd, NULL, NULL, NULL, error)) {
-        g_prefix_error(error, "Failed mounting '%s' to '%s': ",
-                       source, mount_point);
+    cxt = mnt_new_context();
+    if (!cxt) {
+        g_set_error(error, PU_ERROR, PU_ERROR_MOUNT,
+                    "Failed creating mount context while mounting '%s' to '%s'",
+                    source, mount_point);
+        return FALSE;
+    }
+    mnt_context_set_target(cxt, mount_point);
+    mnt_context_set_source(cxt, source);
+    ret = mnt_context_mount(cxt);
+    if (ret || mnt_context_get_status(cxt) != 1) {
+        g_set_error(error, PU_ERROR, PU_ERROR_MOUNT,
+                    "Failed mounting '%s' to '%s'", source, mount_point);
+        mnt_free_context(cxt);
         return FALSE;
     }
 
+    mnt_free_context(cxt);
     return TRUE;
 }
 
 gboolean
 pu_umount(const gchar *mount_point,
           GError **error)
 {
-    g_autofree gchar *cmd = g_strdup_printf("umount %s", mount_point);
+    gint ret;
+    struct libmnt_context *cxt;
```

Comment:
Small nit-pick: I'd prefer naming the context `ctx` instead of `cxt`, as it seems more logical IMHO. :sweat_smile:

Reply:
I took this from https://github.com/util-linux/util-linux/blob/master/sys-utils/mount.c, but I can change it, if this really bothers you.